agent/dns.go

```diff
@@ -429,39 +429,52 @@ RPC:
 	}
 }
 
-// formatNodeRecord takes a Node and returns an A, AAAA, or CNAME record
+// formatTxtRecords takes a kv-map and returns it as "[k=]v" for non-empty k not starting with rfc1035-
+func (d *DNSServer) formatTxtRecords(meta map[string]string) (txt []string) {
+	for k, v := range meta {
+		if strings.HasPrefix(k, "rfc1035-") {
+			txt = append(txt, v)
+		} else {
+			txt = append(txt, k+"="+v)
+		}
+	}
+	return txt
+}
+
+// formatNodeRecord takes a Node and returns an A, AAAA, TXT or CNAME record
 func (d *DNSServer) formatNodeRecord(node *structs.Node, addr, qName string, qType uint16, ttl time.Duration, edns bool) (records []dns.RR) {
 	// Parse the IP
 	ip := net.ParseIP(addr)
 	var ipv4 net.IP
 	if ip != nil {
 		ipv4 = ip.To4()
 	}
+
 	switch {
 	case ipv4 != nil && (qType == dns.TypeANY || qType == dns.TypeA):
-		return []dns.RR{&dns.A{
+		records = append(records, &dns.A{
```

Comment:
We have to append in order to support ANY queries when the node also has a metadata key-values